yente/search/nested.py

```diff
@@ -55,55 +77,145 @@ def nest_entity(
+        inbound_query = make_inbound_query(inbound_ids, prop)
+        if inbound_query:
+            shoulds.append(inbound_query)
+        outbound_query = make_outbound_query(outbound_ids)
+        if outbound_query:
+            shoulds.append(outbound_query)
 
         if not len(shoulds):
             break
+
         query = {
             "bool": {
                 "should": shoulds,
                 "minimum_should_match": 1,
                 "must_not": [{"ids": {"values": list(entities.keys())}}],
             }
         }
-
         resp = await provider.search(
             index=settings.ENTITY_INDEX,
             query=query,
-            size=settings.MAX_RESULTS,
+            size=limit,
+            from_=offset,
+            sort=sort,
         )
-        reverse = []
-        next_entities.clear()
+
+        if total is None:
+            total = result_total(resp)
+        # Prepare for second iteration
+        limit = settings.MAX_RESULTS
+        offset = 0
+        inbound_ids = []
+        outbound_ids.clear()
+
         for adj in result_entities(resp):
             if adj.id is None:
                 continue
             entities[adj.id] = adj
 
-            for prop, value in adj.itervalues():
-                if prop.type != registry.entity:
+            for adj_prop, value in adj.itervalues():
+                if adj_prop.type != registry.entity:
                     continue
                 if adj.schema.edge and value not in entities:
-                    next_entities.add(value)
+                    outbound_ids.add(value)
 
                 inverted.setdefault(value, set())
-                if prop.reverse is not None:
-                    inverted[value].add((prop.reverse, adj.id))
-
-    return nest_entity(root, entities, inverted, set())
+                if adj_prop.reverse is not None:
+                    inverted[value].add((adj_prop.reverse, adj.id))
+
+    truncate_ids = limit or offset
+    nested = nest_entity(root, entities, inverted, set(), truncate_ids)
+    return nested, total
+
+
+async def get_adjacent_entities(
+    provider: SearchProvider, entity: Entity, limit: int, offset: int, sort: List[Any]
+) -> EntityAdjacentResponse:
+    """Queries the requested page of results for each property of type Entity."""
+    tasks = []
+    async with asyncio.TaskGroup() as tg:
+        for prop_name, prop in entity.schema.properties.items():
+            if prop.type != registry.entity:
+                continue
+            task = tg.create_task(
+                get_nested_entity(provider, entity, prop, sort, limit, offset)
+            )
+            tasks.append((prop_name, task))
+    responses = {}
+    for prop_name, task in tasks:
+        prop_response, total = task.result()
+        if total.value:
+            responses[prop_name] = AdjacentResultsResponse(
+                results=prop_response.properties.get(prop_name, []),
```

Comment:
this awkward getting out of the root entity is because nest_entity is recursive. We kinda want an option for it to return the prop, results mapping for the top level. I haven't thought of a nice way to do this yet.